### PR TITLE
feat(plugin): VoiceDownload

### PR DESCRIPTION
**Pull request by puv**
Adds a download button to voice messages.
image

Edit: [061f35a](https://github.com/Vendicated/Vencord/pull/2280/commits/061f35a0db175949ba282e31842017ffd2a5752c) removes usage of a minified variable.
Edit: [02d1838](https://github.com/Vendicated/Vencord/pull/2280/commits/02d18387430959328b40d89dcc4c83f9fc12d148) removes 3 variables that ended up meaningless.
Edit: [1d60852](https://github.com/Vendicated/Vencord/pull/2280/commits/1d608525a95d377c5024a20b49c10c3c7754e6a7) merges constants.ts with updated version.
Edit: [d83e5ab](https://github.com/Vendicated/Vencord/pull/2280/commits/d83e5ab1f2a8a026e3665d8752f0d3aaefa2f008) fixes another constants.ts conflict.

### CODE REVIEW

**Comment by OIRNOIR** [comment]
Vencord needs this imo

File: src/plugins/voiceDownload/index.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+export default definePlugin({
+    name: "VoiceDownload",
+    description: "Download voice messages.",
+    authors: [Devs.puv],
+    patches: [
+        {
+            find: "className:C.rippleContainer",
+            replacement: {
+                match: /\(0,i.jsx\).{0,150},children:.{0,50}\("source",{src:(.{1,2})}\)}\)/,
```
Comment (Vendicated) [inline src/plugins/voiceDownload/index.tsx:21]:
```suggestion
                match: /\(0,\i\.jsx\).{0,150},children:.{0,50}\("source",{src:(\i)}\)}\)/,
```

File: src/plugins/voiceDownload/index.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import { Devs } from "@utils/constants";
+import definePlugin from "@utils/types";
+
+export default definePlugin({
+    name: "VoiceDownload",
+    description: "Download voice messages.",
+    authors: [Devs.puv],
+    patches: [
+        {
+            find: "className:C.rippleContainer",
```
Comment (Vendicated) [inline src/plugins/voiceDownload/index.tsx:19]: don't use minified variables like `C`

File: src/plugins/voiceDownload/index.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+import "./style.css";
+
+import ErrorBoundary from "@components/ErrorBoundary";
+import { Devs } from "@utils/constants";
+import definePlugin from "@utils/types";
+
+export default definePlugin({
+    name: "VoiceDownload",
+    description: "Download voice messages.",
```
Comment (Vendicated) [inline src/plugins/voiceDownload/index.tsx:15]: this description should be more descriptive

File: src/plugins/voiceDownload/index.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            replacement: {
+                match: /\(0,i.jsx\).{0,150},children:.{0,50}\("source",{src:(.{1,2})}\)}\)/,
+                replace: "[$&, $self.renderDownload($1)]"
+            }
+        }
+    ],
+
+    renderDownload: function (src) {
+        return (
+            <ErrorBoundary>
```
Comment (Vendicated) [inline src/plugins/voiceDownload/index.tsx:29]: there's no need to add an ErrorBoundary. there's nothing in your components that could possibly error

File: src/plugins/voiceDownload/index.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        {
+            find: "className:C.rippleContainer",
+            replacement: {
+                match: /\(0,i.jsx\).{0,150},children:.{0,50}\("source",{src:(.{1,2})}\)}\)/,
+                replace: "[$&, $self.renderDownload($1)]"
+            }
+        }
+    ],
+
+    renderDownload: function (src) {
```
Comment (Vendicated) [inline src/plugins/voiceDownload/index.tsx:27]:
```suggestion
    renderDownload(src) {
```

File: src/plugins/voiceDownload/index.tsx (hunk 1 above)
Comment (puv) [reply src/plugins/voiceDownload/index.tsx]: Fixed in [10d545f](https://github.com/Vendicated/Vencord/pull/2280/commits/10d545f88036e8602c304dcd1a5ab76e9e79f8c2)

File: src/plugins/voiceDownload/index.tsx (hunk 2 above)
Comment (puv) [reply src/plugins/voiceDownload/index.tsx]: Fixed in [10d545f](https://github.com/Vendicated/Vencord/pull/2280/commits/10d545f88036e8602c304dcd1a5ab76e9e79f8c2)

File: src/plugins/voiceDownload/index.tsx (hunk 3 above)
Comment (puv) [reply src/plugins/voiceDownload/index.tsx]: Fixed in [10d545f](https://github.com/Vendicated/Vencord/pull/2280/commits/10d545f88036e8602c304dcd1a5ab76e9e79f8c2)

File: src/plugins/voiceDownload/index.tsx (hunk 4 above)
Comment (puv) [reply src/plugins/voiceDownload/index.tsx]: Fixed in [10d545f](https://github.com/Vendicated/Vencord/pull/2280/commits/10d545f88036e8602c304dcd1a5ab76e9e79f8c2)

File: src/plugins/voiceDownload/index.tsx (hunk 5 above)
Comment (puv) [reply src/plugins/voiceDownload/index.tsx]: Fixed in [10d545f](https://github.com/Vendicated/Vencord/pull/2280/commits/10d545f88036e8602c304dcd1a5ab76e9e79f8c2)

**Comment by Golonchy** [comment]
we need this

**Comment by poppcicle** [comment]
Any update on this? I'd really love to see this get merged

**Comment by Vendicated** [comment]
thank you a lot for your contribution!

i made a few changes myself cause i was too lazy to request changes, namely to not hardcode css colours (if you hardcode white it will look awful on light theme) and to prefix css classes

**Pull request merged**